pkg/apis/kudo/v1alpha1/instance_types.go

```diff
@@ -16,65 +16,372 @@ limitations under the License.
+}
+
+// GetPlanInProgress returns plan status of currently active plan or nil if no plan is running
+func (i *Instance) GetPlanInProgress() *PlanStatus {
+	for _, p := range i.Status.PlanStatus {
+		if p.Status.IsRunning() {
+			return &p
+		}
+	}
+	return nil
+}
+
+// NoPlanEverExecuted returns true is this is new instance for which we never executed any plan
+func (i *Instance) NoPlanEverExecuted() bool {
+	for _, p := range i.Status.PlanStatus {
+		if p.Status != ExecutionNeverRun {
+			return false
+		}
+	}
+	return true
+}
+
+// EnsurePlanStatusInitialized initializes plan status for all plans this instance supports
+// it does not trigger run of any plan
+// it either initializes everything for a fresh instance without any status or tries to adjust status after OV was updated
+func (i *Instance) EnsurePlanStatusInitialized(ov *OperatorVersion) {
+	if i.Status.PlanStatus == nil {
+		i.Status.PlanStatus = make(map[string]PlanStatus)
+	}
+
+	for planName, plan := range ov.Spec.Plans {
+		planStatus := &PlanStatus{
+			Name:   planName,
+			Status: ExecutionNeverRun,
+			Phases: make([]PhaseStatus, 0),
+		}
+
+		existingPlanStatus, planExists := i.Status.PlanStatus[planName]
+		if planExists {
+			planStatus.Status = existingPlanStatus.Status
+		}
+		for _, phase := range plan.Phases {
+			phaseStatus := &PhaseStatus{
+				Name:   phase.Name,
+				Status: ExecutionNeverRun,
+				Steps:  make([]StepStatus, 0),
+			}
+			existingPhaseStatus, phaseExists := PhaseStatus{}, false
+			if planExists {
+				for _, oldPhase := range existingPlanStatus.Phases {
+					if phase.Name == oldPhase.Name {
+						existingPhaseStatus = oldPhase
+						phaseExists = true
+						phaseStatus.Status = existingPhaseStatus.Status
+					}
+				}
+			}
+			for _, step := range phase.Steps {
+				stepStatus := StepStatus{
+					Name:   step.Name,
+					Status: ExecutionNeverRun,
+				}
+				if phaseExists {
+					for _, oldStep := range existingPhaseStatus.Steps {
+						if step.Name == oldStep.Name {
+							stepStatus.Status = oldStep.Status
+						}
+					}
+				}
+				phaseStatus.Steps = append(phaseStatus.Steps, stepStatus)
+			}
+			planStatus.Phases = append(planStatus.Phases, *phaseStatus)
+		}
+		i.Status.PlanStatus[planName] = *planStatus
+	}
+}
+
+// StartPlanExecution mark plan as to be executed
+func (i *Instance) StartPlanExecution(planName string, ov *OperatorVersion) error {
+	if i.NoPlanEverExecuted() || isUpgradePlan(planName) {
+		i.EnsurePlanStatusInitialized(ov)
+	}
+
+	// update status of the instance to reflect the newly starting plan
+	notFound := true
+	for n1, v := range i.Status.PlanStatus {
+		if v.Name == planName {
+			// update plan status
+			notFound = false
+			planStatus := i.Status.PlanStatus[n1]
+			planStatus.Status = ExecutionPending
+			for i2, p := range v.Phases {
+				planStatus.Phases[i2].Status = ExecutionPending
+				for i3 := range p.Steps {
+					i.Status.PlanStatus[n1].Phases[i2].Steps[i3].Status = ExecutionPending
```

Comment:
it is unfortunate that i is instance here... and we have i2 and i3 as indexes.   I would encourage expanding i into `instance` in this case. and use `i`, `j`,`k` as indexes

Reply:
If I rename this one I would have to rename all of them in the whole file :/ and I kind of like to use i for it, feels go-ish... I renamed the index to `planIndex` let me know if that helps

Reply:
@alenkacz i've also seen <n>Idx to go shorter in a lot of Go, so you could have:

planIdx, phaseIdx, stepIdx

Reply:
this is a rule/situation where we can favor clarity over conciseness, because as it stands it would be really easy to refactor this, not thinking about indexes, and create a runtime out of bounds error. making those index names look like that embeds the semantic right in the access pattern.